### PR TITLE
feat[contracts]: introduce new L1ChugSplashProxy contract

### LINKED ISSUE #707: Add L1 Proxy Contract with ability to pause during upgrade
**Is your feature request related to a problem? Please describe.**
We need to pause all interactions with our L1 contracts during an upgrade. If we don't then we don't know what crazy interactions might happen if someone sent a transaction mid deploy. 

**Describe the solution you'd like**
Implement a ChugSplashProxy which, before executing any transactions, checks to see if a deployment is currently under-way. This should be called in the fallback function before any interactions with the contract. Also, to enable folks to `eth_call` even during an upgrade, we can use the following logic:

```
function() {
  if (isUpgradeHappening()) {
    require(msg.sender == 0, "Only eth_call is allowed during an upgrade!");
  }
}
```

**Blockers**
https://github.com/ethereum-optimism/optimism/issues/703

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Introduces the `L1ChugSplashProxy` contract. Uses some fancy footwork to avoid function signature collisions with the implementation contract, would appreciate a good amount of review on this.

**TODO**
- [x] Add tests
- [x] Add ability to pause the contract

**Metadata**
- Fixes #707 
- Fixes OP-625

### CODE REVIEW

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (maurelian) [inline packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:58]: is the second `msg.sender == address(0)` possible on L1?

Comment (smartcontracts) [reply packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:58]: Nope. Only possible if you do an `eth_call`.

File: packages/contracts/test/contracts/chugsplash/L1ChugSplashProxy.spec.ts
Comment (maurelian) [inline packages/contracts/test/contracts/chugsplash/L1ChugSplashProxy.spec.ts:123]: Uber nit, but clearer IMO:
```suggestion
    it('should not change the implementation address if the code does not change', async () => {
```

File: packages/contracts/test/contracts/chugsplash/L1ChugSplashProxy.spec.ts
Comment (maurelian) [inline packages/contracts/test/contracts/chugsplash/L1ChugSplashProxy.spec.ts:141]: How about:
```suggestion
    it('should revert if implementation is not set', async () => {
```

File: packages/contracts/test/contracts/chugsplash/L1ChugSplashProxy.spec.ts
Comment (maurelian) [inline packages/contracts/test/contracts/chugsplash/L1ChugSplashProxy.spec.ts:105]: A slightly better test (here and in the other 'pass through' tests) might be to ensure that `setCode` the implementation is called with `bytes4(keccak256("setStorage(bytes32)"))`.

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (maurelian) [inline packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:111]: Since the Bridge will handle ETH deposits, we'll need this to be payable.

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (maurelian) [inline packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:320]: Just based on this comment from the header, I'm curious if this code is mostly lifted from another well tested proxy?

> Very close to being a normal proxy but has added functions `setCode` and `setStorage` for changing the code or storage of the contract

**Review by maurelian — CHANGES_REQUESTED** [review]

**Comment by maurelian** [comment]
Also needs a changeset :)

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (smartcontracts) [reply packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:320]: Not lifted from a known proxy

**Comment by tynes** [comment]
What else is required to get this over the line?

**Comment by smartcontracts** [comment]
> What else is required to get this over the line?

Needs review from 2 of @ben-chain @elenadimitrova @maurelian

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (ben-chain) [inline packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:133]: This function does not appear to check that the deployment is successful; either by checking the `_code.length` or by checking that `newImplementation != address(0)`.  Do we want to do that?

File: packages/contracts/test/contracts/chugsplash/L1ChugSplashProxy.spec.ts
Comment (ben-chain) [inline packages/contracts/test/contracts/chugsplash/L1ChugSplashProxy.spec.ts:68]: should we add a `from: ZERO_ADDRESS` check on the getter here?

**Review by ben-chain — COMMENTED** [review]
A couple questions, but this looks pretty much good to go to me.

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (smartcontracts) [reply packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:133]: I think that's a good idea.

File: packages/contracts/test/contracts/chugsplash/L1ChugSplashProxy.spec.ts
Comment (smartcontracts) [reply packages/contracts/test/contracts/chugsplash/L1ChugSplashProxy.spec.ts:68]: Can do just for the sake of having it.

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (maurelian) [reply packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:328]: OK, I compared it to OZ's [proxy.sol](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/Proxy.sol#L15-L44), and it's functionally identical, so that's nice

**Review by maurelian — APPROVED** [review]

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (karlfloersch) [inline packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:31]: Definitely not a requirement but I imagine `immutable` would be useful here to express these constants explicitly as they are derived. That's a nit though - this seems like a fine approach.

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (karlfloersch) [inline packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:160]: 0x20 puts the pointer passed the abi length prefix right?

**Review by karlfloersch — COMMENTED** [review]
This is extremely dope & looks good to me as well

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (smartcontracts) [reply packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:160]: Correct

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (smartcontracts) [reply packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:31]: I'll give it a shot

Comment (smartcontracts) [reply packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:31]: Seems like that totally works. Nice!

Comment (smartcontracts) [reply packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:31]: Welp nevermind, this sort of works but it creates some headaches because we use `OWNER_KEY` in the constructor.

Comment (smartcontracts) [reply packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:31]: So I'm not gonna do it.

**Comment by tynes** [comment]
The system is still called `ChugSplash` which isn't particularly descriptive in what it does. We spoke about renaming it before actually using it and if this is merged then its unlikely that it will ever be renamed.

**Comment by smartcontracts** [comment]
> The system is still called `ChugSplash` which isn't particularly descriptive in what it does. We spoke about renaming it before actually using it and if this is merged then its unlikely that it will ever be renamed.

I'm open to name changes

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (ben-chain) [inline packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:8]:
```suggestion
 * @dev Basic ChugSplash proxy contract for L1. Very close to being a normal delegate proxy but has added
```

File: packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol
Comment (ben-chain) [inline packages/contracts/contracts/chugsplash/L1ChugSplashProxy.sol:162]: I think this could be triggered if the `_code.length >= 24000`.  I like doing it this way instead of explicitly checking for that case though; it's a better catch-all.  But we should update this comment accordingly.

**Review by ben-chain — APPROVED** [review]
This LGTM, just a couple of comment requests :)

**Review by karlfloersch — APPROVED** [review]
Alright I did another once over. The only thing that I could think of as a risk for forward compatibility was adding events for `setCode`, `setStorage`, and `_setOwner`. I don't now *why* we might want these events, but I figure there's not much cost to adding them. **However**, I am not strongly opinionated here and therefore will approve the PR even in its current state. I will also approve if you decide to add events to the stateful operations.

Annnyway, again, this is awesome and LGTM!

**Pull request merged**
